File: renderers/AGenUI/platforms/ios/AGenUI/Classes/AGenUI.swift
//
//  AGenUI.swift
//  AGenUI
//
//  Created by acoder-ai-infra on 2026/3/18.
//

import Foundation

/// AGenUI SDK global entry point
///
/// Singleton responsible for global registration of theme, DesignToken, day/night mode, skills, components, and image loaders.
@objc(AGenUISDK) public class AGenUISDK: NSObject {

    private override init() {
        super.init()
    }

    // MARK: - Engine Bridge

    private static var engineBridge: AGenUIEngineBridge {
        return AGenUIEngineBridge.sharedInstance()
    }

    // MARK: - Global Configuration

    // MARK: - Theme Management

    /// Register default theme configuration
    ///
    /// Registers theme and design token configurations to customize UI component appearance.
    ///
    /// - Parameters:
    ///   - theme: Theme configuration JSON string
    ///   - designToken: Design token configuration JSON string
    /// - Returns: AGenUIError with result and message fields
    @objc public static func registerDefaultTheme(_ theme: String, designToken: String) -> AGenUIError {
        Logger.shared.debug("registerDefaultTheme - theme length: \(theme.count), designToken length: \(designToken.count)")
        
        // 1. Wrap theme with "default" key
        guard let themeData = theme.data(using: .utf8),
              let themeObject = try? JSONSerialization.jsonObject(with: themeData) else {
            Logger.shared.error("Failed to parse theme JSON")
            return AGenUIError(result: false, message: "Invalid theme JSON format")
        }
        
        let wrappedTheme: [String: Any] = ["default": themeObject]
        guard let wrappedData = try? JSONSerialization.data(withJSONObject: wrappedTheme),
              let wrappedThemeJson = String(data: wrappedData, encoding: .utf8) else {
            Logger.shared.error("Failed to serialize wrapped theme JSON")
            return AGenUIError(result: false, message: "Failed to wrap theme JSON")
        }
        
        // 2. Register theme
        let themeResult = engineBridge.loadThemeConfig(wrappedThemeJson)
        if !themeResult {
            Logger.shared.error("Theme registration failed")
            return AGenUIError(result: false, message: "Theme registration failed")
        }
        
        // 3. Register DesignToken
        let designTokenResult = engineBridge.loadDesignTokenConfig(designToken)
        if !designTokenResult {
            Logger.shared.error("DesignToken registration failed")
            return AGenUIError(result: false, message: "DesignToken registration failed")
        }
        
        Logger.shared.info("Default theme registered successfully")
        return AGenUIError(result: true, message: "Success")
    }

    /// Set day/night mode
    ///
    /// - Parameter mode: "light" or "dark"
    @objc public static func setDayNightMode(_ mode: String) {
        Logger.shared.debug("setDayNightMode: \(mode)")
        guard mode == "light" || mode == "dark" else {
            Logger.shared.warning("setDayNightMode: invalid mode '\(mode)', expected 'light' or 'dark'")
            return
        }
        engineBridge.setDayNightMode(mode)
        Logger.shared.info("Day/Night mode set to: \(mode)")
    }

    // MARK: - Component Registration

    /// Register a custom component factory
    ///
    /// - Parameters:
    ///   - type: Component type name
    ///   - creator: Factory closure to create the component
    /// - Note: Overwrites if the component type already exists
    @objc public static func registerComponent(_ type: String, creator: @escaping (String, [String: Any]) -> Component) {
        Logger.shared.debug("registerComponent: \(type)")
        ComponentRegister.shared.register(type, creator: creator)
        Logger.shared.info("Custom component registered: \(type)")
    }
    
    @objc public static func unRegisterComponent(_ type: String) {
        Logger.shared.debug("unregisterComponent: \(type)")
        ComponentRegister.shared.unregister(type)
        Logger.shared.info("Custom component unregistered: \(type)")
    }

    // MARK: - Image Loader Registration

    /// Register a global image loader
    ///
    /// - Parameter loader: A loader instance implementing the ImageLoader protocol
    @objc(registerImageLoader:)
    public static func registerImageLoader(_ loader: ImageLoader) {
        Logger.shared.debug("registerImageLoader: \(type(of: loader))")
        ImageLoaderConfiguration.shared.loader = loader
        Logger.shared.info("ImageLoader registered: \(type(of: loader))")
    }
    
    // MARK: - FunctionCall Management

    /// Register a platform function with the engine
    ///
    /// - Parameter function: Object implementing the AGenUIFunctionProtocol
    @objc(registerFunction:)
    public static func registerFunction(_ function: Function) {
        let config = function.functionConfig
        let name = config.name
        guard !name.isEmpty else {
            Logger.shared.error("registerFunction: function name is empty")
            return
        }
        
        let configJson = config.toJSON()
        Logger.shared.debug("registerFunction - name: \(name)")
        
        // Create bridge callback that delegates to the function instance
        let bridgeCallback: AGenUIFunctionCallCallback = { [weak function] argsJson in
            
            guard let function = function else {
                return "{\"status\":\"Error\",\"error\":\"Function deallocated\"}"
            }
            
            let result = function.execute(argsJson)
            
            let resultDict: [String: Any] = [
                "status": result.result ? "Success" : "Error",
                "data": result.value
            ]
            
            guard let resultData = try? JSONSerialization.data(withJSONObject: resultDict),
                  let resultJson = String(data: resultData, encoding: .utf8) else {
                return "{\"status\":\"Error\",\"error\":\"Failed to serialize result\"}"
            }
            
            return resultJson
        }
        
        engineBridge.registerFunction(name, config: configJson, callback: bridgeCallback)
        Logger.shared.info("FunctionCall registered: \(name)")
    }

    /// Unregister a previously registered platform function
    ///
    /// - Parameter name: Function name to unregister
    @objc(unregisterFunctionWithName:)
    public static func unregisterFunction(_ name: String) {
        guard !name.isEmpty else {
            Logger.shared.error("unregisterFunction: name is empty")
            return
        }
        
        engineBridge.unregisterFunction(name)
        Logger.shared.info("FunctionCall unregistered: \(name)")
    }

    // MARK: - Version Info

    /// Get the AGenUI SDK version
    ///
    /// - Returns: Version string
    @objc public static func getVersion() -> String {
        return "2.0.0"
    }
}

// MARK: - AGenUIError

@objc public class AGenUIError: NSObject {
    /// Whether the operation succeeded
    @objc public let result: Bool
    /// Error message
    @objc public let message: String
    
    @objc public init(result: Bool, message: String) {
        self.result = result
        self.message = message
    }
}

File: renderers/AGenUI/platforms/android/src/main/java/com/amap/agenui/AGenUI.java
package com.amap.agenui;

import android.content.Context;
import android.content.res.AssetManager;
import android.util.Log;

import androidx.annotation.Keep;
import androidx.annotation.RestrictTo;

import com.amap.agenui.function.IFunction;
import com.amap.agenui.function.PlatformFunction;
import com.amap.agenui.render.component.ComponentRegistry;
import com.amap.agenui.render.component.IComponentFactory;
import com.amap.agenui.render.image.ImageLoader;
import com.amap.agenui.render.image.ImageLoaderConfig;
import com.amap.agenui.render.surface.ThemeException;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;

@Keep
public class AGenUI {
    private static final String TAG = "AGenUI";

    static {
        System.loadLibrary("amap_AGenUI");
    }

    private static volatile AGenUI sInstance = null;
    private static final Object sLock = new Object();

    private static final String SDK_VERSION = "2.0.0";

    private volatile long nativePtr = 0;
    private volatile boolean isInitialized = false;
    private volatile Context appContext = null;

    /**
     * Private constructor to prevent direct external instantiation
     */
    private AGenUI() {
    }

    /**
     * Returns the AGenUI singleton instance
     *
     * @return AGenUI singleton instance
     */
    public static AGenUI getInstance() {
        if (sInstance == null) {
            synchronized (sLock) {
                if (sInstance == null) {
                    sInstance = new AGenUI();
                }
            }
        }
        return sInstance;
    }

    /**
     * Initializes the AGenUI Engine
     *
     * Performs the following steps:
     * 1. Loads Native modules
     * 2. Creates the Engine instance (initAGenUIEngine)
     * 3. Sets the working directory
     * 4. Copies template files to the sandbox
     * 5. Initializes SkillManager and registers platform Skills
     *
     * @throws RuntimeException if initialization fails
     */
    public void initialize(Context applicationContext) {
        synchronized (sLock) {
            if (isInitialized) {
                Log.w(TAG, "Module already initialized");
                return;
            }

            try {
                appContext = applicationContext.getApplicationContext();
                nativePtr = nativeInitAGenUIEngine();
                Log.i(TAG, "AGenUI Engine created: nativePtr=" + nativePtr);
                nativeSetWorkdir(getInternalFilesDir());
                copyAssetsToSandbox();
                ComponentRegistry.registerBuiltInComponents();

                isInitialized = true;
                Log.i(TAG, "AGenUI Engine initialized successfully");
            } catch (Exception e) {
                Log.e(TAG, "Failed to initialize AGenUI Engine", e);
                throw new RuntimeException("Failed to initialize AGenUI Engine", e);
            }
        }
    }

    /**
     * Checks whether the Engine has been initialized
     *
     * @return true if the Engine is initialized
     */
    public boolean isInitialized() {
        return isInitialized && nativePtr != 0;
    }


    /**
     * Creates a SurfaceManager instance
     *
     * @return engineId (instance identifier)
     * @throws IllegalStateException if the Engine is not initialized or the native layer fails to create one
     */
    public int createSurfaceManager() throws IllegalStateException {
        if (!isInitialized()) {
            throw new IllegalStateException("createSurfaceManager: AGenUI engine is not initialized");
        }
        int engineId = nativeCreateSurfaceManager();
        if (engineId == 0) {
            throw new IllegalStateException("createSurfaceManager: native call failed");
        }
        Log.i(TAG, "SurfaceManager created: engineId=" + engineId);
        return engineId;
    }

    /**
     * Destroys a SurfaceManager instance
     *
     * @param engineId The engineId of the SurfaceManager to destroy
     */
    public void destroySurfaceManager(int engineId) {
        if (!isInitialized()) {
            Log.w(TAG, "destroySurfaceManager: Engine not initialized");
            return;
        }
        nativeDestroySurfaceManager(engineId);
        Log.i(TAG, "SurfaceManager destroyed: engineId=" + engineId);
    }


    private boolean isConfigValid(String methodName, String config) {
        if (config == null || config.isEmpty()) {
            Log.w(TAG, methodName + ": config is null or empty");
            return false;
        }
        if (!isInitialized()) {
            Log.e(TAG, methodName + ": Engine not initialized");
            return false;
        }
        return true;
    }

    private boolean loadThemeConfig(String themeConfig) {
        if (!isConfigValid("loadThemeConfig", themeConfig)) {
            return false;
        }
        return nativeLoadThemeConfig(themeConfig);
    }

    private boolean loadDesignTokenConfig(String designTokenConfig) {
        if (!isConfigValid("loadDesignTokenConfig", designTokenConfig)) {
            return false;
        }
        return nativeLoadDesignTokenConfig(designTokenConfig);
    }

    /**
     * Sets the day/night mode
     *
     * @param mode Mode value: "light" or "dark"
     */
    public void setDayNightMode(String mode) {
        if (mode == null || mode.isEmpty()) {
            Log.w(TAG, "setDayNightMode: mode is null or empty");
            return;
        }
        if (!isInitialized()) {
            Log.e(TAG, "setDayNightMode: Engine not initialized");
            return;
        }
        nativeSetDayNightMode(mode);
    }

    /**
     * Registers the default theme configuration
     * <p>
     * Registers both the theme JSON and DesignToken JSON simultaneously.
     *
     * @param theme       Theme configuration JSON string
     * @param designToken DesignToken configuration JSON string
     * @throws ThemeException if registration fails
     */
    public void registerDefaultTheme(String theme, String designToken) throws ThemeException {
        boolean themeOk = loadThemeConfig(theme);
        if (!themeOk) {
            throw new ThemeException("Failed to register theme config");
        }
        boolean tokenOk = loadDesignTokenConfig(designToken);
        if (!tokenOk) {
            throw new ThemeException("Failed to register design token config");
        }
        Log.i(TAG, "✓ Default theme registered successfully");
    }


    public void registerFunction(IFunction function) {
        nativeRegisterFunction(
                function.getConfig().getName(),
                function.getConfig().toJSON(),
                new PlatformFunction(function));
    }

    public void unregisterFunction(String name) {
        nativeUnregisterFunction(name);
    }


    /**
     * Registers a custom component factory
     * <p>
     * If the component type already exists, it will be overwritten. Takes effect immediately
     * after registration and is shared across all Surfaces.
     *
     * @param type    Component type (e.g. "MyCustomCard")
     * @param creator Component factory instance
     */
    public void registerComponent(String type, IComponentFactory creator) {
        if (type == null || type.isEmpty() || creator == null) {
            Log.w(TAG, "registerComponent: invalid parameters");
            return;
        }
        ComponentRegistry.registerComponent(type, creator);
        Log.i(TAG, "registerComponent: type=" + type);
    }

    public void unregisterComponent(String type) {
        if (type == null || type.isEmpty()) {
            Log.w(TAG, "unregisterComponent: invalid parameters");
            return;
        }
        ComponentRegistry.unregisterComponent(type);
        Log.i(TAG, "unregisterComponent: type=" + type);
    }


    /**
     * Registers a global image loader
     * <p>
     * All image components will use this loader to load network images.
     *
     * @param loader ImageLoader instance
     */
    public void registerImageLoader(ImageLoader loader) {
        if (loader == null) {
            Log.w(TAG, "registerImageLoader: loader is null");
            return;
        }
        ImageLoaderConfig.getInstance().setLoader(loader);
        Log.i(TAG, "registerImageLoader: success");
    }


    /**
     * Returns the AGenUI SDK version number
     *
     * @return SDK version number
     */
    public static String getVersion() {
        return SDK_VERSION;
    }


    /**
     * Returns the Native Engine pointer
     *
     * @return Native Engine pointer
     * @throws IllegalStateException if the Engine is not initialized
     */
    public long getNativePtr() {
        if (!isInitialized()) {
            throw new IllegalStateException("Engine not initialized");
        }
        return nativePtr;
    }

    /**
     * Destroys the Engine and releases all Native resources.
     * This method should be called when the application exits.
     */
    public void destroy() {
        synchronized (sLock) {
            if (!isInitialized) {
                Log.w(TAG, "Engine not initialized, nothing to destroy");
                return;
            }

            try {
                if (nativePtr != 0) {
                    nativeDestroyAGenUIEngine();
                    Log.i(TAG, "Engine destroyed successfully");
                }
            } catch (Exception e) {
                Log.e(TAG, "Error destroying engine", e);
            } finally {
                nativePtr = 0;
                isInitialized = false;
                sInstance = null;
            }
        }
    }

    @RestrictTo(androidx.annotation.RestrictTo.Scope.LIBRARY)
    public String getInternalFilesDir() {
        if (appContext == null) {
            throw new IllegalStateException("AGenUI not initialized");
        }
        return appContext.getFilesDir().getAbsolutePath();
    }


    /**
     * Copies template files from assets/agenui to the sandbox ui_templates directory
     */
    private void copyAssetsToSandbox() {
        try {
            String targetDirPath = getInternalFilesDir() + File.separator + "data" + File.separator + "ui_templates";
            File targetDir = new File(targetDirPath);
            if (!targetDir.exists() && !targetDir.mkdirs()) {
                Log.e(TAG, "copyAssetsToSandbox: Failed to create directory: " + targetDirPath);
                return;
            }

            AssetManager assetManager = appContext.getAssets();
            String[] files = assetManager.list("agenui");
            if (files == null || files.length == 0) {
                Log.w(TAG, "copyAssetsToSandbox: No files found in assets/agenui");
                return;
            }

            for (String fileName : files) {
                File targetFile = new File(targetDir, fileName);
                if (targetFile.exists()) {
                    continue;
                }
                InputStream is = null;
                FileOutputStream fos = null;
                try {
                    is = assetManager.open("agenui" + File.separator + fileName);
                    fos = new FileOutputStream(targetFile);
                    byte[] buffer = new byte[4096];
                    int bytesRead;
                    while ((bytesRead = is.read(buffer)) != -1) {
                        fos.write(buffer, 0, bytesRead);
                    }
                    fos.flush();
                } catch (IOException e) {
                    Log.e(TAG, "copyAssetsToSandbox: Failed to copy file: " + fileName, e);
                } finally {
                    if (is != null) {
                        try { is.close(); } catch (IOException ignored) {}
                    }
                    if (fos != null) {
                        try { fos.close(); } catch (IOException ignored) {}
                    }
                }
            }
            Log.i(TAG, "copyAssetsToSandbox: Templates copied to " + targetDirPath);
        } catch (Exception e) {
            Log.e(TAG, "copyAssetsToSandbox: Failed to copy assets", e);
        }
    }


    private native long nativeInitAGenUIEngine();
    private native void nativeDestroyAGenUIEngine();

    public static native int nativeCreateSurfaceManager();
    public static native void nativeDestroySurfaceManager(int engineId);

    public static native void nativeSetWorkdir(String workdir);

    private static native boolean nativeLoadThemeConfig(String themeConfig);
    private static native boolean nativeLoadDesignTokenConfig(String designTokenConfig);
    private static native void nativeSetDayNightMode(String mode);

    public static native void nativeRegisterFunction(String name, String config, Object function);
    public static native void nativeUnregisterFunction(String name);
    public static native void nativeOnAsyncCallbackResult(long callbackPtr, int status, String data, String error);
}
